CCpuInfo split, rework, and cleanup

lrusak opened the pull request:
I'm PRing this early to get some feedback about some things. I'm also hoping some platform maintainers will jump in to help with the split.

~~This depends on PR #16097 so please ignore that commit here.~~

The point of this PR is to use inheritance instead of ifdefs to split the `CCPUInfo` class.

How can we bring the class up without a global? Is using `CServiceBroker` the best way?

How can we create the class in `CApplication`? is it ok to ifdef for each platform or should we finally move to implement a platform based init?

I have tested this on x86 and arm, it should also probably be tested on aarch64.

I have removed some feature and some may have to be added, for example

Removed:
 - per cpu model name, speed, etc

Needs adding:
 - bogomips (is this really needed on modern linux kernels?)
 - more thermal sysfs paths for linux

I've also fixed the test case but I'm not sure if what I did is the best way. Please advise.

Todo:
 - [x] android
 - [x] osx
 - [x] ios
 - [x] win32
 - [x] win10
 - [x] freebsd

lrusak committed:
CCPUInfo: split platforms and rework

afedchin commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp:
```diff
@@ -1735,8 +1735,7 @@ bool CAMLCodec::OpenDecoder(CDVDStreamInfo &hints)
 
-  std::string strScaler;
-  SysfsUtils::GetString("/sys/class/ppmgr/ppscaler", strScaler);
+  std::string strScaler = SysfsUtils::GetString("/sys/class/ppmgr/ppscaler");
```
unrelated changes

afedchin commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp:
```diff
@@ -1786,8 +1785,7 @@ bool CAMLCodec::OpenAmlVideo(const CDVDStreamInfo &hints)
 {
-  std::string vfmMap;
-  SysfsUtils::GetString("/sys/class/vfm/map", vfmMap);
+  std::string vfmMap = SysfsUtils::GetString("/sys/class/vfm/map");
```
unrelated changes

afedchin commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp:
```diff
@@ -1855,8 +1853,7 @@ void CAMLCodec::Reset()
   // set the system blackout_policy to leave the last frame showing
-  int blackout_policy;
-  SysfsUtils::GetInt("/sys/class/video/blackout_policy", blackout_policy);
+  int blackout_policy = SysfsUtils::GetInt("/sys/class/video/blackout_policy");
```
unrelated changes

afedchin commented on xbmc/platform/linux/RBP.h:
```diff
@@ -83,7 +83,7 @@ class CRBP
   int GetGpuMem() { return m_gpu_mem; }
   bool GetCodecMpg2() { return m_codec_mpg2_enabled; }
-  int RaspberryPiVersion() { return g_cpuInfo.getCPUCount() == 1 ? 1 : 2; };
+  int RaspberryPiVersion() { return CServiceBroker::GetCPUInfo()->GetCPUCount() == 1 ? 1 : 2; };
```
move to cpp file. header must be as simple as possible

afedchin commented on xbmc/utils/SystemInfo.cpp:
```diff
@@ -304,7 +304,7 @@ std::string CSysInfoJob::GetBatteryLevel()
 double CSysInfoJob::GetCPUFrequency()
 {
 #if defined (TARGET_POSIX) || defined(TARGET_WINDOWS)
```
unneeded ifdefs. GetCPUFrequency should return 0 depending on implementation

afedchin commented on xbmc/utils/SystemInfo.cpp:
```diff
@@ -491,39 +491,39 @@ bool CSysInfo::GetDiskSpace(std::string drive,int& iTotal, int& iTotalFree, int&
 
 std::string CSysInfo::GetCPUModel()
```
I see no reason to keep these "proxy" methods

afedchin commented:
I see no windows/uwp implementations. at least c&p from the original.

lrusak commented:
> I see no windows/uwp implementations. at least c&p from the original.

@afedchin I added win32 and win10, please fix them as needed.

As for the other platforms I just added dummy headers.
@peak3d for android
@Memphiz for ios/osx

lrusak committed:
CSystemInfo: cleanup duplicate methods

Memphiz commented:
So the old lmplementation from OS X and iOS is removed and currently the dummies are there only - did I get this right?

lrusak commented:
> So the old lmplementation from OS X and iOS is removed and currently the dummies are there only - did I get this right?

Currently yes. Until I or someone else has time to work on it.

fuzzard commented:
For better or worse, ios currently defines TARGET_LINUX which causes failures on build. IOS also HAS_NEON which causes the below failure

```
cpuinfo/xbmc/platform/linux/CPUInfoLinux.cpp:23:10: fatal error: 
      'asm/hwcap.h' file not found
#include <asm/hwcap.h>
         ^~~~~~~~~~~~~
1 error generated.
```

On top of that ios doesnt have cpuid.h

```
In file included from /cpuinfo/xbmc/platform/linux/CPUInfoLinux.cpp:26:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/10.0.1/include/cpuid.h:25:2: error: 
      this header is for x86 only
#error this header is for x86 only
 ^
```

I dont have an answer for you just yet, but may need to ifdef !defined(TARGET_DARWIN_IOS) some things. Im doing a build to test a fix at the moment, so ill come back with some info a bit later i hope.

Obviously not having ios/darwin pull from platform/linux would be ideal, but i imagine thats a far greater task than is desired right now.

fuzzard commented on xbmc/platform/linux/CPUInfoLinux.cpp:
```diff
@@ -0,0 +1,312 @@
+
+float CCPUInfoLinux::GetCPUFrequency()
+{
+  int value;
```
compiler warnings for an uninitialized variable and what happens with the following if being false. either need to init variable, or just remove the if if it can never fail i guess so there isnt a possibility of a bad divide operation

```
[ 16%] Building CXX object build/platform/linux/CMakeFiles/linuxsupport.dir/CPUInfoLinux.cpp.o
/cpuinfo/xbmc/platform/linux/CPUInfoLinux.cpp:266:7: warning: 
      variable 'value' is used uninitialized whenever 'if' condition is false
      [-Wsometimes-uninitialized]
  if (SysfsUtils::Has("/sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq"))
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/cpuinfo/xbmc/platform/linux/CPUInfoLinux.cpp:269:3: note: 
      uninitialized use occurs here
  value /= 1000.0;
  ^~~~~

```

fuzzard commented on xbmc/platform/linux/CPUInfoLinux.cpp:
```diff
@@ -0,0 +1,312 @@
+  temperature.SetValid(false);
+
+  int value{-1};
+  char scale;
```
uninitialized variable, just throws a warning.

lrusak commented:
> Obviously not having ios/darwin pull from platform/linux would be ideal, but i imagine thats a far greater task than is desired right now.

This will have to be solved for this PR. The sole purpose of this PR is to breakup the platforms into platform specific code. `CCPUInfoLinux` is now specific to linux. I can change the cmake code to only build this for the linux windowing platforms to satisfy this for now but that would need to be fixed in the near future.

Any Darwin code that uses `TARGET_LINUX` should be changed to use it's own code. If there is code that is shared between linux and darwin that code can be moved to platform/posix

pkerling commented on xbmc/ServiceBroker.h:
```diff
@@ -131,6 +132,10 @@ class CServiceBroker
   static void RegisterDecoderFilterManager(CDecoderFilterManager* manager);
   static CDecoderFilterManager* GetDecoderFilterManager();
 
+  static std::shared_ptr<CCPUInfo> GetCPUInfo();
```
Why make this a shared pointer? Does some component need to use the cpu info for an extended period of time so that it might vanish during use? Or is this to prevent races on shutdown?

lrusak replied:
I haven't tested this yet to be honest. I think originally I was just worried about multiple locations requesting the pointer at once.

pkerling commented on xbmc/platform/linux/CPUInfoLinux.cpp:
```diff
@@ -0,0 +1,312 @@
+
+namespace
+{
+enum cpuStates
```
```suggestion
enum CpuStates
```

pkerling commented on xbmc/platform/linux/CPUInfoLinux.cpp:
```diff
@@ -0,0 +1,312 @@
+
+  for (std::size_t core = 0; core < cpuData.size(); core++)
+  {
+    auto activeTime = cpuData.at(core).GetActiveTime() - m_cores.at(core).m_activeTime;
```
```suggestion
    auto activeTime = cpuData[core].GetActiveTime() - m_cores[core].m_activeTime;
```

Why use `.at()` here?

garbear replied:
`.at()` can be seen as safer, as it will throw an exception instead of accessing invalid memory.

pkerling replied:
I know, but

1. other places don't really use it, only here for some reason
2. the for loop already makes sure that only valid indices are used

pkerling commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,231 @@
+  if (GetSystemTimes(&idleTime, &kernelTime, &userTime) == 0)
+    return false;
+
+  idle = (uint64_t(idleTime.dwHighDateTime) << 32) + uint64_t(idleTime.dwLowDateTime);
```
Time to clean up these function-style casts?

pkerling commented on xbmc/utils/CPUInfo.cpp:
```diff
@@ -6,1032 +6,54 @@
-#endif
-
-  return true;
+  return std::move(coreInfo);
```
wat?

pkerling commented on xbmc/utils/CPUInfo.h:
```diff
@@ -8,125 +8,108 @@
-#define CPU_FEATURE_3DNOWEXT 1 << 9
-#define CPU_FEATURE_ALTIVEC  1 << 10
-#define CPU_FEATURE_NEON     1 << 11
+enum CPU_FEATURE_ENUM
```
```suggestion
enum CpuFeature
```

pkerling commented on xbmc/Application.cpp:
```diff
@@ -362,6 +392,38 @@ void CApplication::Preflight()
 {
+// hack
+#if defined(TARGET_LINUX)
+  std::shared_ptr<CCPUInfo> cpuInfo = std::make_shared<CCPUInfoLinux>();
```
Move all of this to a factory, IMO that is enough

pkerling commented on xbmc/utils/CPUInfo.h:
```diff
@@ -8,125 +8,108 @@
-  std::string  m_cpuSerial;
-  int          m_cpuCount;
+  std::string m_cpuVendor{"N/A"};
+  std::string m_cpuModel{"N/A"};;
```
```suggestion
  std::string m_cpuModel{"N/A"};
```

pkerling commented:
> bogomips (is this really needed on modern linux kernels?)

don't think anyone seriously needs this

lrusak commented:
is `CPlatform*::Init()` a good place to do registration?

lrusak commented:
jenkins build this please

lrusak commented:
I've updated all the platforms now. They still need fixing though as I don't build those other platforms.

I'm waiting for #16113 to go in but otherwise I think this should be good to start reviewing.

@Rechi how can I get jenkins to build this PR now?

fuzzard commented on xbmc/platform/darwin/ios/CPUInfoIos.cpp:
```diff
@@ -0,0 +1,75 @@
+
+#include "CPUInfoIos.h"
+
+#include "platform/linux/LinuxResourceCounter.h"
```
platform/posix/PosixResourceCounter.h

from the rebase

fuzzard commented on xbmc/platform/darwin/ios/CPUInfoIos.cpp:
```diff
@@ -0,0 +1,75 @@
+}
+
+CCPUInfoIos::CCPUInfoIos()
+  : m_resourceCounter(new CLinuxResourceCounter())
```
CPosixResourceCounter

fuzzard commented on xbmc/platform/darwin/ios/CPUInfoIos.h:
```diff
@@ -0,0 +1,30 @@
+
+#include <memory>
+
+class CLinuxResourceCounter;
```
CPosixResourceCounter

fuzzard commented on xbmc/platform/darwin/ios/CPUInfoIos.h:
```diff
@@ -0,0 +1,30 @@
+  bool GetTemperature(CTemperature& temperature) override { return false; }
+
+private:
+  std::unique_ptr<CLinuxResourceCounter> m_resourceCounter;
```
CPosixResourceCounter

fuzzard commented on xbmc/platform/darwin/osx/CPUInfoOsx.cpp:
```diff
@@ -0,0 +1,116 @@
+}
+
+CCPUInfoOsx::CCPUInfoOsx()
+  : m_resourceCounter(new CLinuxResourceCounter())
```
Header include just above and CPosixResourceCounter

fuzzard commented on xbmc/platform/darwin/osx/CPUInfoOsx.h:
```diff
@@ -0,0 +1,27 @@
+  bool GetTemperature(CTemperature& temperature) override { return false; }
+
+private:
+  std::unique_ptr<CLinuxResourceCounter> m_resourceCounter;
```
Class above and CPosixResourceCounter

lrusak commented:
currently waiting on #16210

Rechi commented on xbmc/platform/darwin/ios/CPUInfoIos.cpp:
```diff
@@ -0,0 +1,76 @@
+#include <string>
+
+#include <sys/types.h>
+#include <sys/sysctl.h>
```
wrong order

Rechi commented on xbmc/platform/darwin/osx/CPUInfoOsx.cpp:
```diff
@@ -0,0 +1,124 @@
+#include <string>
+
+#include <sys/types.h>
+#include <sys/sysctl.h>
```
wrong order

Rechi commented on xbmc/platform/freebsd/CPUInfoFreebsd.cpp:
```diff
@@ -0,0 +1,151 @@
+
+#include <sys/types.h>
+#include <sys/sysctl.h>
+#include <sys/resource.h>
```
wrong order

Rechi commented on xbmc/platform/linux/RBP.h:
```diff
@@ -20,7 +20,8 @@
 #include "DllBCM.h"
 #include "OMXCore.h"
-#include "xbmc/utils/CPUInfo.h"
+#include "utils/CPUInfo.h"
```
wrong order

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,241 @@
+#include "utils/SysfsUtils.h"
+#include "utils/Temperature.h"
+
+#include "platform/win32/CharsetConverter.h"
```
this is in the wrong section

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,241 @@
+          (cnt.CStatus == PDH_CSTATUS_VALID_DATA || cnt.CStatus == PDH_CSTATUS_NEW_DATA))
+      {
+        const LONGLONG coreTotal = cnt.SecondValue,
+                       coreIdle  = cnt.FirstValue;
```
make this a separate declaration

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,241 @@
+        const LONGLONG coreTotal = cnt.SecondValue,
+                       coreIdle  = cnt.FirstValue;
+        const LONGLONG deltaTotal = coreTotal - core.m_totalTime,
+                       deltaIdle  = coreIdle - core.m_idleTime;
```
make this a separate declaration

Rechi commented on xbmc/platform/android/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+set(HEADERS CPUInfoAndroid.cpp)
```
```suggestion
set(SOURCES CPUInfoAndroid.cpp)
```

lrusak commented:
Someone else will have to fix the windows changes. I cannot do it.

lrusak commented:
@Rechi or @afedchin can one of you take on fixing the windows changes?

DaveTBlake commented:
Having a look at fixing these changes for Windows...
Made several edits and got it compiling, but I have run into an issue with using `CServiceBroker` in WinMain.cpp to check if SSE2 is supported. `CServiceBroker::GetCPUInfo()` is null at that point as it has not been registered. Not sure how to get around that or where `CPlatform::Init()` can be called.

Rechi commented on xbmc/platform/android/CPUInfoAndroid.cpp:
```diff
@@ -0,0 +1,32 @@
+#include "CPUInfoAndroid.h"
+
+#include "platform/android/activity/AndroidFeatures.h"
+#include "utils/Temperature.h"
```
wrong include order

lrusak replied:
how so?

Rechi replied:
https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#7-headers

lrusak replied:
be specific

Rechi replied:
- Own header file: `CPUInfoAndroid.h`
- Other Kodi includes (platform independent): `utils/Temperature.h`
- Other Kodi includes (platform specific): `platform/android/activity/AndroidFeatures.h`

lrusak replied:
why do we want that?

Rechi commented on xbmc/platform/darwin/ios/CPUInfoIos.cpp:
```diff
@@ -0,0 +1,76 @@
+#include "CPUInfoIos.h"
+
+#include "platform/posix/PosixResourceCounter.h"
+#include "utils/Temperature.h"
```
wrong include order

Rechi commented on xbmc/platform/darwin/ios/CPUInfoIos.cpp:
```diff
@@ -0,0 +1,76 @@
+
+#include <sys/sysctl.h>
+#include <sys/types.h>
+
```
remove the empty line

Rechi commented on xbmc/platform/darwin/osx/CPUInfoOsx.cpp:
```diff
@@ -0,0 +1,124 @@
+
+#include "platform/darwin/osx/smc.h"
+#include "platform/posix/PosixResourceCounter.h"
+#include "utils/Temperature.h"
```
wrong include order

Rechi commented on xbmc/platform/win10/CPUInfoWin10.cpp:
```diff
@@ -0,0 +1,127 @@
+
+  if (HasNeon())
+    m_cpuFeatures |= CPU_FEATURE_NEON;
+
```
remove the empty line

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,283 @@
+#include "CPUInfoWin32.h"
+
+#include "ServiceBroker.h"
+#include "platform/win32/CharsetConverter.h"
```
wrong include order

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,283 @@
+    m_cpuQueryLoad = nullptr;
+
+
+
```
one or two empty lines are enough

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,283 @@
+          (cnt.CStatus == PDH_CSTATUS_VALID_DATA || cnt.CStatus == PDH_CSTATUS_NEW_DATA))
+      {
+        const LONGLONG coreTotal = cnt.SecondValue;
+        const LONGLONG coreIdle  = cnt.FirstValue;
```
one space is enough

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,283 @@
+        const LONGLONG coreTotal = cnt.SecondValue;
+        const LONGLONG coreIdle  = cnt.FirstValue;
+        const LONGLONG deltaTotal = coreTotal - core.m_totalTime;
+        const LONGLONG deltaIdle  = coreIdle - core.m_idleTime;
```
one space is enough

Rechi commented on xbmc/utils/SystemInfo.cpp:
```diff
@@ -301,7 +301,7 @@ std::string CSysInfoJob::GetBatteryLevel()
 {
 #if defined (TARGET_POSIX) || defined(TARGET_WINDOWS)
-  return double (g_cpuInfo.getCPUFrequency());
+  return double (CServiceBroker::GetCPUInfo()->GetCPUFrequency());
```
no space between `double` and `(`

Rechi commented on xbmc/platform/freebsd/CPUInfoFreebsd.cpp:
```diff
@@ -0,0 +1,149 @@
+#include "CPUInfoFreebsd.h"
+
+#include "utils/log.h"
+#include "utils/Temperature.h"
```
wrong include order

lrusak replied:
how?

lrusak replied:
@Rechi it's really hard to push this PR through if you don't answer my questions

Rechi replied:
> Included header files have to be sorted (case sensitive) alphabetically [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#7-headers

Rechi replied:
still wrong

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,282 @@
+
+#include <intrin.h>
+#include <Pdh.h>
+#include <PdhMsg.h>
```
wrong include order

lrusak replied:
how?

Rechi replied:
still wrong

Rechi commented on xbmc/platform/win32/WinMain.cpp:
```diff
@@ -10,6 +10,7 @@
 #include "CompileInfo.h"
 #include "platform/Environment.h"
 #include "platform/xbmc.h"
+#include "ServiceBroker.h"
```
wrong include order

lrusak replied:
how?

Rechi replied:
still wrong

Rechi reviewed (commented):
The `TestCPUInfo.GetUsedPercentage` test segfaulted two times on FreeBSD.

lrusak committed:
win32: remove check for sse2 during init

lrusak commented:
I added a commit to remove the SSE2 check on windows during startup. It seems that not even windows 7 supports non-SSE2 CPU's so we don't need to double check this. I hope this is an ok solution @DaveTBlake @afedchin 

https://www.ghacks.net/2018/06/21/windows-7-support-dropped-for-cpus-without-sse2/

Rechi commented on xbmc/Application.cpp:
```diff
@@ -474,11 +474,11 @@ bool CApplication::Create(const CAppParamParser &params)
+  std::string cpuModel(CServiceBroker::GetCPUInfo()->GetCPUModel());
   if (!cpuModel.empty())
-    CLog::Log(LOGNOTICE, "Host CPU: %s, %d core%s available", cpuModel.c_str(), g_cpuInfo.getCPUCount(), (g_cpuInfo.getCPUCount() == 1) ? "" : "s");
+    CLog::Log(LOGNOTICE, "Host CPU: %s, %d core%s available", cpuModel.c_str(), CServiceBroker::GetCPUInfo()->GetCPUCount(), (CServiceBroker::GetCPUInfo()->GetCPUCount() == 1) ? "" : "s");
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/Application.cpp:
```diff
@@ -474,11 +474,11 @@ bool CApplication::Create(const CAppParamParser &params)
+    CLog::Log(LOGNOTICE, "Host CPU: %s, %d core%s available", cpuModel.c_str(), CServiceBroker::GetCPUInfo()->GetCPUCount(), (CServiceBroker::GetCPUInfo()->GetCPUCount() == 1) ? "" : "s");
   else
-    CLog::Log(LOGNOTICE, "%d CPU core%s available", g_cpuInfo.getCPUCount(), (g_cpuInfo.getCPUCount() == 1) ? "" : "s");
+    CLog::Log(LOGNOTICE, "%d CPU core%s available", CServiceBroker::GetCPUInfo()->GetCPUCount(), (CServiceBroker::GetCPUInfo()->GetCPUCount() == 1) ? "" : "s");
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/guilib/guiinfo/SystemGUIInfo.cpp:
```diff
@@ -283,7 +283,7 @@ bool CSystemGUIInfo::GetLabel(std::string& value, const CFileItem *item, int con
     }
     case SYSTEM_GET_CORE_USAGE:
-      value = StringUtils::Format("%4.2f", g_cpuInfo.GetCoreInfo(std::atoi(info.GetData3().c_str())).m_fPct);
+      value = StringUtils::Format("%4.2f", CServiceBroker::GetCPUInfo()->GetCoreInfo(std::atoi(info.GetData3().c_str())).m_usagePercent);
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/platform/darwin/ios/CPUInfoIos.cpp:
```diff
@@ -0,0 +1,76 @@
+}
+
+CCPUInfoIos::CCPUInfoIos()
+  : m_resourceCounter(new CPosixResourceCounter())
```
> For lines up to [line length](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#line-length) everything stays on one line, excluding the braces which must be on the following lines.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#116-constructor-initialzation-lists

Rechi commented on xbmc/platform/darwin/osx/CPUInfoOsx.cpp:
```diff
@@ -0,0 +1,125 @@
+}
+
+CCPUInfoOsx::CCPUInfoOsx()
+  : m_resourceCounter(new CPosixResourceCounter())
```
> For lines up to [line length](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#line-length) everything stays on one line, excluding the braces which must be on the following lines.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#116-constructor-initialzation-lists

Rechi commented on xbmc/platform/win10/CPUInfoWin10.cpp:
```diff
@@ -0,0 +1,117 @@
+  if (!m_nextUsedReadTime.IsTimePast())
+    return m_lastUsedPercentage;
+
+  if (!winrt::Windows::Foundation::Metadata::ApiInformation::IsTypePresent(L"Windows.System.Diagnostics.SystemDiagnosticInfo"))
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/platform/linux/CPUInfoLinux.cpp:
```diff
@@ -0,0 +1,322 @@
+
+  return true;
+}
+
```
remove the empty line

Rechi commented on xbmc/platform/win10/CPUInfoWin10.cpp:
```diff
@@ -0,0 +1,117 @@
+    return 0;
+  }
+
+  auto diagnostic = winrt::Windows::System::Diagnostics::SystemDiagnosticInfo::GetForCurrentSystem();
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,282 @@
+
+  HKEY hKeyCpuRoot;
+
+  if (RegOpenKeyExW(HKEY_LOCAL_MACHINE, L"HARDWARE\\DESCRIPTION\\System\\CentralProcessor", 0, KEY_READ, &hKeyCpuRoot) == ERROR_SUCCESS)
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,282 @@
+    DWORD num = 0;
+    wchar_t subKeyName[200]; // more than enough
+    DWORD subKeyNameLen = sizeof(subKeyName) / sizeof(wchar_t);
+    while (RegEnumKeyExW(hKeyCpuRoot, num++, subKeyName, &subKeyNameLen, nullptr, nullptr, nullptr, nullptr) == ERROR_SUCCESS)
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,282 @@
+        DWORD bufSize = sizeof(buf);
+        DWORD valType;
+        if (!modelfound &&
+            RegQueryValueExW(hCpuKey, L"ProcessorNameString", nullptr, &valType, LPBYTE(buf), &bufSize) == ERROR_SUCCESS &&
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,282 @@
+            valType == REG_SZ)
+        {
+          m_cpuModel = FromW(buf, bufSize / sizeof(wchar_t));
+          m_cpuModel = m_cpuModel.substr(0, m_cpuModel.find(char(0))); // remove extra null terminations
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,282 @@
+        }
+        bufSize = sizeof(buf);
+        if (!vendorfound &&
+            RegQueryValueExW(hCpuKey, L"VendorIdentifier", nullptr, &valType, LPBYTE(buf), &bufSize) == ERROR_SUCCESS &&
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,282 @@
+            valType == REG_SZ)
+        {
+          m_cpuVendor = FromW(buf, bufSize / sizeof(wchar_t));
+          m_cpuVendor = m_cpuVendor.substr(0, m_cpuVendor.find(char(0))); // remove extra null terminations
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,282 @@
+
+  if (PdhOpenQueryW(nullptr, 0, &m_cpuQueryFreq) == ERROR_SUCCESS)
+  {
+    if (PdhAddEnglishCounterW(m_cpuQueryFreq, L"\\Processor Information(0,0)\\Processor Frequency", 0, &m_cpuFreqCounter) != ERROR_SUCCESS)
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,282 @@
+  size_t idleTime = 0;
+  size_t totalTime = 0;
+
+  idleTime = (static_cast<size_t>(idleTimestamp.dwHighDateTime) << 32) + static_cast<size_t>(idleTimestamp.dwLowDateTime);
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,282 @@
+  size_t totalTime = 0;
+
+  idleTime = (static_cast<size_t>(idleTimestamp.dwHighDateTime) << 32) + static_cast<size_t>(idleTimestamp.dwLowDateTime);
+  kernelTime = (static_cast<size_t>(kernelTimestamp.dwHighDateTime) << 32) + static_cast<size_t>(kernelTimestamp.dwLowDateTime);
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,282 @@
+
+  idleTime = (static_cast<size_t>(idleTimestamp.dwHighDateTime) << 32) + static_cast<size_t>(idleTimestamp.dwLowDateTime);
+  kernelTime = (static_cast<size_t>(kernelTimestamp.dwHighDateTime) << 32) + static_cast<size_t>(kernelTimestamp.dwLowDateTime);
+  userTime = (static_cast<size_t>(userTimestamp.dwHighDateTime) << 32) + static_cast<size_t>(userTimestamp.dwLowDateTime);
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,282 @@
+          core.m_usagePercent = 0;
+        else
+          core.m_usagePercent = load;
+        if (load >= 0 || deltaTotal > 5 * 10 * 1000 * 1000) // do not update (smooth) values for 5 seconds on negative loads
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/utils/SystemInfo.cpp:
```diff
@@ -251,7 +251,7 @@ bool CSysInfoJob::DoWork()
   m_info.internetState     = GetInternetState();
   m_info.videoEncoder      = GetVideoEncoder();
-  m_info.cpuFrequency      = GetCPUFreqInfo();
+  m_info.cpuFrequency      = StringUtils::Format("%4.0f MHz", CServiceBroker::GetCPUInfo()->GetCPUFrequency());
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

> Do not use whitespace to vertically align around operators or values. This causes problems on code review if one needs to realign all values to their new position, producing unnecessarily large diffs.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#35-vertical-alignment

Rechi commented on xbmc/Application.cpp:
```diff
@@ -474,13 +474,16 @@ bool CApplication::Create(const CAppParamParser &params)
+              (CServiceBroker::GetCPUInfo()->GetCPUCount() == 1) ? "" : "s");
 
-  //! @todo - move to CPlatformXXX ???
+    //! @todo - move to CPlatformXXX ???
```
indentation is wrong

lrusak replied:
this is what clang-format wants

Rechi replied:
I know, but it is wrong.

pkerling replied:
IMO we should not manually edit whatever clang-format outputs (didn't see the original change in context though and it's not there any more)

Rechi replied:
It just had indented this single line, not a whole block.

DaveTBlake commented:
>I added a commit to remove the SSE2 check on windows during startup. ... I hope this is an ok solution 

Sorry I have no idea @lrusak, although it seems reasonable. No Win7 here to test, but I'm sure we still have users out there with Kodi on Win7  (old laptop as HTPC etc.)

lrusak commented:
> > I added a commit to remove the SSE2 check on windows during startup. ... I hope this is an ok solution
> 
> Sorry I have no idea @lrusak, although it seems reasonable. No Win7 here to test, but I'm sure we still have users out there with Kodi on Win7 (old laptop as HTPC etc.)

Yes I'm not dropping support for windows 7. The last processors without SSE2 are pentium 3. Surely it's not possible to run Kodi (or run it well) on such a processor.

DaveTBlake commented:
Personally I think that is fine @lrusak , I'm just also aware I am no expert! 
But please push this over the line before there are more cosmetic changes or rebases :)

lrusak commented:
jenkins build this please

lrusak commented:
ok to merge?

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

lrusak commented:
looks like the freebsd build is segfaulting when the `TestCPUInfo` test is running. I have no way to debug that.

rozhuk-im commented:
@lrusak I can try to debug this, copy-paste how to build and run test will help me with that, if you can provide it.

lrusak commented:
> @lrusak I can try to debug this, copy-paste how to build and run test will help me with that, if you can provide it.

You can just fetch my branch and build according to the docs here -> https://github.com/xbmc/xbmc/blob/master/docs/README.FreeBSD.md

I imagine kodi won't even run if the test itself segfaults (which basically is just a minimal kodi environment).

rozhuk-im commented:
Thanks!

```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  CCPUInfoFreebsd::CCPUInfoFreebsd (this=0x81426e018) at /home/rim/tmp/111/xbmc/xbmc/platform/freebsd/CPUInfoFreebsd.cpp:61
61	    m_cores[core.m_id] = core;
[Current thread is 1 (LWP 100638)]
warning: File "/usr/local/lib/libpython3.6m.so.1.0-gdb.py" auto-loading has been declined by your `auto-load safe-path' set to "$debugdir:$datadir/auto-load".
To enable execution of this file add
	add-auto-load-safe-path /usr/local/lib/libpython3.6m.so.1.0-gdb.py
line to your configuration file "/root/.gdbinit".
--Type <RET> for more, q to quit, c to continue without paging--bt
To completely disable this security protection add
	set auto-load safe-path /
line to your configuration file "/root/.gdbinit".
For more information about this security protection see the
"Auto-loading safe path" section in the GDB manual.  E.g., run from the shell:
	info "(gdb)Auto-loading safe path"
(gdb) bt
#0  CCPUInfoFreebsd::CCPUInfoFreebsd (this=0x81426e018) at /home/rim/tmp/111/xbmc/xbmc/platform/freebsd/CPUInfoFreebsd.cpp:61
#1  0x0000000000c80a57 in std::__1::__compressed_pair_elem<CCPUInfoFreebsd, 1, false>::__compressed_pair_elem (this=<optimized out>) at /usr/include/c++/v1/memory:2140
#2  std::__1::__compressed_pair<std::__1::allocator<CCPUInfoFreebsd>, CCPUInfoFreebsd>::__compressed_pair<std::__1::allocator<CCPUInfoFreebsd>, true> (this=<optimized out>, __t=...)
    at /usr/include/c++/v1/memory:2241
#3  std::__1::__shared_ptr_emplace<CCPUInfoFreebsd, std::__1::allocator<CCPUInfoFreebsd> >::__shared_ptr_emplace (this=0x81426e000, __a=...) at /usr/include/c++/v1/memory:3667
#4  std::__1::shared_ptr<CCPUInfoFreebsd>::make_shared<>() () at /usr/include/c++/v1/memory:4331
#5  std::__1::make_shared<CCPUInfoFreebsd> () at /usr/include/c++/v1/memory:4710
#6  CCPUInfo::GetCPUInfo () at /home/rim/tmp/111/xbmc/xbmc/platform/freebsd/CPUInfoFreebsd.cpp:40
#7  0x0000000001041e97 in CPlatform::Init (this=<optimized out>) at /home/rim/tmp/111/xbmc/xbmc/platform/Platform.cpp:32
#8  0x0000000000c6a3c9 in CPlatformPosix::Init (this=0x8143745e0) at /home/rim/tmp/111/xbmc/xbmc/platform/posix/PlatformPosix.cpp:15
#9  0x000000000134ff63 in CServiceManager::InitStageOne (this=0x8142620e0) at /home/rim/tmp/111/xbmc/xbmc/ServiceManager.cpp:90
#10 0x00000000012c5d6a in CApplication::Create (this=0x814256000, params=...) at /home/rim/tmp/111/xbmc/xbmc/Application.cpp:377
#11 0x0000000001041f56 in XBMC_Run (renderGUI=<optimized out>, params=...) at /home/rim/tmp/111/xbmc/xbmc/platform/xbmc.cpp:33
#12 0x0000000000ad23a6 in main (argc=<optimized out>, argv=<optimized out>) at /home/rim/tmp/111/xbmc/xbmc/platform/posix/main.cpp:77
```

rozhuk-im commented:
Build and run OK on FreeBSD 12.1 amd64.

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

lrusak commented:
@rozhuk-im can you please check this test on FreeBSD?
```
15:33:48 [  FAILED  ] TestCPUInfo.GetTemperature
```

rozhuk-im commented:
Probably this happen because: ```bool GetTemperature(CTemperature& temperature) override { return false; }```
I see no implementation ```GetTemperature()``` for FreeBSD and some other platforms.
My implementation #16779, but it will fails test too, because test server does not load or do not have temp sensors in kernel.

lrusak commented:
jenkins build this please

lrusak commented:
Jenkins build this please

Rechi commented:
This PR broke Windows UWP.
```
xbmc/utils/CPUInfo.cp(75): error C3861: 'popen': identifier not found
xbmc/utils/CPUInfo.cp(79): error C3861: 'pclose': identifier not found
```

Rechi commented on xbmc/platform/win32/CPUInfoWin32.cpp:
```diff
@@ -0,0 +1,289 @@
+  else
+    m_cpuQueryLoad = nullptr;
+
+#ifndef _M_ARM
```
This ifdef existed for win10 not for win32 in the original code.

Rechi commented on xbmc/utils/test/TestCPUInfo.cpp:
```diff
@@ -22,109 +22,80 @@
+TEST_F(TestCPUInfo, GetTemperature)
 {
-  TemporarySetting command(CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_cpuTempCmd, "echo '50 c'");
+  CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_cpuTempCmd = "echo '50 c'";
```
This test fails on windows.

Rechi commented on xbmc/utils/CPUInfo.cpp:
```diff
@@ -6,1032 +6,89 @@
   else if (scale == 'F' || scale == 'f')
     temperature = CTemperature::CreateFromFahrenheit(value);
-  else
-    return false;
```
Because this else was removed, every scale results in a valid temperature.
